// orchestrate-mongo/test/util.js
var async = require('async');
var OrchestrateMongo = require('../lib-cov');
var mongo = require('mongodb');
var orchestrate = require('orchestrate');

exports.setup = function (done) {
  this.collection = process.env.MONGODB_COLLECTION;
  this.database = process.env.MONGODB_DATABASE;
  this.host = process.env.MONGODB_HOST || 'localhost';
  this.port = process.env.MONGODB_PORT || 27017;
  this.orc = orchestrate(process.env.ORCHESTRATE_API_KEY);
  this.watcher = OrchestrateMongo({ 
    mongodb: {
      database: this.database
    },
    collection: this.collection
  });

  this.db = new mongo.Db(this.database, new mongo.Server(this.host, this.port), { 
    w: 1,
    slaveOk: true
  });

  var self = this;
  async.waterfall([
    function (done) {
      self.db.open(done);
    },
    function (db, done) {
      self.mongo_db = db;
      self.mongo_db.createCollection(self.collection, done);
    },
    function (coll, done) {
      self.mongo_collection = coll;
      done();
    }
  ], done);
};
